aggregator/list.go

```diff
@@ -94,22 +103,23 @@ type metricList struct {
 	log           log.Logger
 	timeLock      *sync.RWMutex
 	maxFlushSize  int
+	partitionFn   PartitionFn
 	flushHandler  Handler
 	encoderPool   msgpack.BufferedEncoderPool
 	resolution    time.Duration
 	flushInterval time.Duration
 	flushMgr      FlushManager
 
-	aggregations       *list.List
-	lastFlushedNanos   int64
-	encoder            msgpack.AggregatedEncoder
-	toCollect          []*list.Element
-	closed             bool
-	encodeFn           encodeFn
-	flushBeforeFn      flushBeforeFn
-	consumeAggMetricFn aggMetricFn
-	discardAggMetricFn aggMetricFn
-	metrics            metricListMetrics
+	closed              bool
+	aggregations        *list.List
+	lastFlushedNanos    int64
+	encodersByPartition []msgpack.AggregatedEncoder
```

Comment:
Just a thought (admittedly could be totally off the mark here): would it make sense for the list to be aware of the partition ranges instead of the total number of partitions? In this way, `encodersByPartition` would be much smaller and we would be able to more easily support a much higher number of partitions. Also, since the handlers are per partition range, it would be consistent with what they expect.

Reply:
That seems to cause tighter coupling between the partitioned handler and the list in that now the list needs to decide which partition range a metric belongs to. The interface would also look a bit weird (e.g., the `Handle` function would take a partition range instead of a partition).

I'm not too concerned about the overhead to keep encoders per partition, given 16 shards per instance, 1K partitions, 1.44KB max flush size, the space taken by the encoders is 16 * 1K * 1.44KB = 20MB, which is quite reasonable.

Reply:
Sounds good, wasn't totally sold on the idea, just figured I would just play devil's advocate.